### ISSUE TITLE
Proposal: `where` keyword

### ISSUE THREAD

**Issue by arthur-xavier** [elm-lang/elm-compiler#1389]
tl;dr write complex functions focusing on the function itself, by using the `where` keyword instead of `let ... in`, like this:

```elm
f x = x + y / z
  where
    y = x*2
    z = 1 + y
```

# Motivation
When reading Elm code, the one thing that most bothers me (and I guess many people also) is the fact that the actual code of a complex function lies down below from where one would search (which is right after the `=`), such as in:

```elm
calc =
  let
    (three, four) = (3, 4)
    hypotenuse a b = sqrt (a^2 + b^2)
  in
    hypotenuse three four
```

# Proposed solution

My proposal is to have a `where` keyword which would behave just as Haskell's one:

```elm
calc = hypotenuse three four
  where
    (three, four) = (3, 4)
    hypotenuse a b = sqrt (a^2 + b^2)
```

The main advantage of this solution against the current syntax is that it allows a much quicker reading and understanding of a function.

**Comment by arthur-xavier** [comment]
I'll close this since I just found out I overlooked the biggest discussion of this century on `where` vs `let ... in` on #621 😅
